### ISSUE TITLE
Rule "no-useless-concat" reports false positive

### ISSUE THREAD

**Issue by steffenweber** [eslint/eslint#3575]
ESLint version:
```
v1.3.0
```

Source code:

```javascript
var number = 10;
var string = (number + 1) + 'px';
window.alert(string); // eslint-disable-line no-alert
```

ESLint output:
```
2:27  error  Unexpected string concatenation of literals  no-useless-concat
```

If I'm not mistaken then this concatenation is not useless because `number` is not a string.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by Dashed** [comment]
I think it's because it can be written as:

```js
var string = 11 + 'px';
window.alert(string); // eslint-disable-line no-alert
```

**Comment by Steffen185** [comment]
That's only true for the simple test case.

The following code causes the same error and cannot be changed in this way:

```
function foo(number) {
	var string = (number + 1) + 'px';
	window.alert(string);
}
```

**Comment by Dashed** [comment]
@Steffen185 you're totally right. I outlined the issue in this code:

```js
const number = 10;

// we want to do this because of type evaluation order
console.log(number + 1 + 'px'); // => 11px

// what no-useless-concat thinks we want to do; but not really
console.log(number + '1px'); // => 101px

// note the early string type coercion
console.log('' + number + 1 + 'px'); // => 101px
```

-----

@hzoo Type coercion order must be considered. See above.

**Comment by hzoo** [comment]
It would be easier to just not error on different types at all although that would not error on the simpler cases of `1 + '1'`?

**Comment by Dashed** [comment]
@hzoo I think it can be simpler than trying all the types. 

Suppose new rule implementation is called `no-useless-concat_new`. And let current implementation be called `no-useless-concat_old`.
For string concatenation, `A_0 ... A_n`, if you can find `A_j` that is a literal string (`0 <= j < n`), and subexpression `A_(j+1) .. A_n` satisfies `no-useless-concat_old` rule, then `no-useless-concat_new` is satisfied.

**Comment by Dashed** [comment]
I don't know how to iterate the AST in eslint, but I think this can be adapted https://github.com/eslint/eslint/blob/master/lib/rules/prefer-template.js

**Comment by mysticatea** [comment]
@hzoo @Dashed I'm thankful for your PRs. But could I work on this?
I want to tweak both `prefer-template` and `no-useless-concat` together.
Related: https://github.com/eslint/eslint/issues/3589#issuecomment-136089452

**Comment by Dashed** [comment]
@mysticatea yep go ahead :+1:

**Comment by TheSavior** [comment]
I just ran into this on our codebase as well. Glad to see you guys are working on this.
